hippogriff12/Asta/BJS2-96257-projects

Charly-vision opened the pull request.

Charly-vision committed:
-functionality for creating, editing, and searching for projects has been implemented
на проверку
(пока без unit-тестов)

Charly-vision committed:
-functionality for creating, editing, and searching for projects has been implemented
на проверку
(пока без unit-тестов)

Charly-vision committed:
Merge remote-tracking branch 'origin/hippogriff12/Asta/BJS2-96257-projects' into hippogriff12/Asta/BJS2-96257-projects

Charly-vision committed:
-made changes
-covered the classes with unit-tests

Scherbi1 commented on src/main/java/faang/school/projectservice/dto/client/project/ProjectCreateDto.java:
```diff
@@ -0,0 +1,14 @@
+        @NotBlank
+        @Size(max = 128)
+        String name,
+        @Size(max = 4096)
```
а тут проверку @NotBlank или не нужна?

Scherbi1 commented on src/main/java/faang/school/projectservice/mapper/ProjectMapper.java:
```diff
@@ -0,0 +1,49 @@
+                .status(dto.status())
+                .visibility(dto.visibility())
+                .createdAt(dto.createdAt())
+                .updatedAt(dto.updatedAt())
```
а если какое то поле null придет из дто?

Scherbi1 commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,95 @@
+    private final ProjectMapper projectMapper;
+    private final ProjectValidator projectValidator;
+
+    @Transactional
```
почему @Transactional написала?

Scherbi1 commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,68 @@
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public ProjectDto createProject(Long ownerId,
```
ownerId откуда берется? не забыла @PathVariable?

Scherbi1 commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,68 @@
+    public List<ProjectDto> getProjectsByFilter(
+            @RequestParam(required = false) String name,
+            @RequestParam(required = false) String status,
+            @RequestHeader("x-user-id") Long userId) {
```
@RequestHeader("x-user-id") Long userId точно так надо?

20cognosce replied:
Да, выглядит не очень в общем стиле, у нас есть юзер контекст)

Scherbi1 commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,68 @@
+    public List<ProjectDto> getAllProjects() {
+        log.info("Get all projects request");
+        return projectFacade.getAll();
+    }
```
ты в createProject использовала @ResponseStatus(HttpStatus.CREATED) почему не использовала его везде?
По типу @ResponseStatus(HttpStatus.ок  (что то такое)

Scherbi1 commented on src/main/java/faang/school/projectservice/validator/project/ProjectValidator.java:
```diff
@@ -0,0 +1,49 @@
+
+    public void validateUniqueProjectNameForOwner(String projectName, Long ownerId) {
+        if (projectRepository.existsByOwnerIdAndName(ownerId, projectName)) {
+            throw new IllegalArgumentException("Project with this name already exists for user");
```
я бы кастомные ошибки создавал

Scherbi1 commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,163 @@
+                .visibility(ProjectVisibility.PUBLIC)
+                .createdAt(LocalDateTime.now())
+                .updatedAt(LocalDateTime.now())
+                .build();
```
тут точно так много полей надо? половина из них разве не должна быть присвоена в create?
ты в create   поля присваиваешь:
        project.setOwnerId(ownerId);
        project.setStatus(ProjectStatus.CREATED);
         
        А тут ты их заранее присвоила в бефоре, не стоит так делать
        лишнего не надо делать

Scherbi1 commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,95 @@
+        project.setStatus(ProjectStatus.CREATED);
+        LocalDateTime now = LocalDateTime.now();
+        project.setCreatedAt(now);
+        project.setUpdatedAt(now);
```
у тебя в entity автоматически проставляется время

mipkasnow commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,68 @@
+            @RequestHeader("x-user-id") Long userId) {
+        log.info("Filter projects request: name='{}', status='{}'", name, status);
+
+        ProjectStatus projectStatus = status != null ? ProjectStatus.valueOf(status) : null;
```
Лучше так не делать, enum в @RequestParam надо передавать как enum. 
@RequestParam(required = false) ProjectStatus status - вот так.
Тогда и тернарник не нужен

mipkasnow commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @GetMapping("/{id}")
+    public ProjectDto getProjectById(@PathVariable Long id, Long userId) {
```
Long userId - не помечено никакой аннотацией. Это PathVariable, RequestParam или хедер. Спринг не поймет откуда его брать

mipkasnow commented on src/main/java/faang/school/projectservice/dto/client/project/ProjectUpdateDto.java:
```diff
@@ -0,0 +1,9 @@
+import faang.school.projectservice.model.ProjectVisibility;
+
+public record ProjectUpdateDto(
+        String description,
+        String status,
+        ProjectVisibility visibility
```
Валидация не нужна?

mipkasnow commented on src/main/java/faang/school/projectservice/mapper/ProjectMapper.java:
```diff
@@ -0,0 +1,49 @@
+        if (dto == null) return null;
+
+        return Project.builder()
+                .id(dto.id())
```
сомнения, что надо id проставлять entity. За это же БД отвечает

mipkasnow commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,95 @@
+    @Transactional
+    public ProjectDto updateProject(Long id, ProjectDto dto) {
+        Project project = projectRepository.findById(id)
+                .orElseThrow(() -> new IllegalArgumentException("Project not found"));
```
Давай дефолтный метод напишем в репозитории для поиска любых сущностей, чтобы руками не обрабатывать эксепшены

mipkasnow commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,95 @@
+        Optional.ofNullable(dto.status()).ifPresent(project::setStatus);
+        Optional.ofNullable(dto.description()).ifPresent(project::setDescription);
+
+        project.setUpdatedAt(LocalDateTime.now());
```
это делается самой БД в entity

mipkasnow commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,95 @@
+
+    public ProjectDto getProjectById(Long id, Long userId) {
+        Project project = projectRepository.findById(id)
+                .orElseThrow(() -> new IllegalArgumentException("Project not found"));
```
дефолтный метод в репозитории

mipkasnow commented on src/main/java/faang/school/projectservice/validator/project/ProjectValidator.java:
```diff
@@ -0,0 +1,49 @@
+@Component
+public class ProjectValidator {
+
+    private final ProjectRepository projectRepository;
```
Стоп - мы же договорились, что валидаторы должны быть максимально простыми классами:
1) Не должны быть бинами, должны иметь статические методы. Чтобы не лепить лишние зависимости в сервисе
2) НЕ ДОЛЖНЫ ходить в базу данных => не должны иметь в зависимости репозиторий. Это будет сложно тестировать

mipkasnow commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,163 @@
+    @Mock
+    private ProjectMapper projectMapper;
+    @Mock
+    private ProjectValidator projectValidator;
```
Вот это не нужно было бы, если бы валидатор был простым классом со статик методам

mipkasnow commented on src/main/java/faang/school/projectservice/mapper/ProjectMapper.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.List;
+
+@Component
+public class ProjectMapper {
```
А может лучше сделать все эти методы статическими? тогда и тестировать проще будет, не надо лишнюю зависимость тянуть к себе и мокать это все? ты же тут чисто руками все перекладываешь

mipkasnow commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,163 @@
+    @Mock
+    private ProjectRepository projectRepository;
+    @Mock
+    private ProjectMapper projectMapper;
```
и вот это бы статикой сделать, тогда и мокать не надо

mipkasnow commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,163 @@
+    void createProject_success() {
+        Long ownerId = 10L;
+
+        when(projectMapper.toEntity(projectDto)).thenReturn(project);
```
Вот наглядно пример - маппер у тебя не статика - лишняя зависимость, вынуждена мокировать его поведение, в итоге легко пропустишь ошибку, если в самом маппере у тебя косяк. Делай статикой

mipkasnow commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,163 @@
+        Long ownerId = 10L;
+
+        when(projectMapper.toEntity(projectDto)).thenReturn(project);
+        when(projectRepository.save(any(Project.class))).thenReturn(project);
```
Еще бы через ArgumentCaptor проверить, какой объект в итоге сохраняет репозиторий

mipkasnow commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    @Test
+    void updateProject_success() {
```
1L, description, status -  можно в константу, повторяется 4 раза

mipkasnow commented on src/test/java/faang/school/projectservice/service/project/ProjectServiceTest.java:
```diff
@@ -0,0 +1,163 @@
+
+        assertThatThrownBy(() -> projectService.getProjectById(999L, 10L))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("Project not found");
```
Можно TestUtils утащить для этого из userService, меньше кода писать

mipkasnow commented on src/test/java/faang/school/projectservice/validator/project/ProjectValidatorTest.java:
```diff
@@ -0,0 +1,121 @@
+    void setup() {
+        projectRepository = mock(ProjectRepository.class);
+        validator = new ProjectValidator(projectRepository);
+    }
```
Через @Mock и @InjectMocks же можно сделать то же самое.
Но опять таки - класс лучше сделать статикой и без зависимости репозитория

mipkasnow commented on src/test/java/faang/school/projectservice/validator/project/ProjectValidatorTest.java:
```diff
@@ -0,0 +1,121 @@
+
+        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class,
+                () -> validator.validateUniqueProjectNameForOwner("TestProject", 1L));
+        assertEquals("Project with this name already exists for user", ex.getMessage());
```
TestUtils из userService

mipkasnow commented on src/test/java/faang/school/projectservice/validator/project/ProjectValidatorTest.java:
```diff
@@ -0,0 +1,121 @@
+
+    @Test
+    void validateProjectExists_Passes_WhenProjectNotNull() {
+        Project project = mock(Project.class);
```
Зачем мокать сущность? через new же тот же эффект

mipkasnow commented on src/test/java/faang/school/projectservice/validator/project/ProjectValidatorTest.java:
```diff
@@ -0,0 +1,121 @@
+
+    @Test
+    void validateAccessToProject_Throws_WhenPrivateAndUserNotParticipant() {
+        Project project = mock(Project.class);
```
Сущности не мокаем, а создаем

mipkasnow commented on src/test/java/faang/school/projectservice/validator/project/ProjectValidatorTest.java:
```diff
@@ -0,0 +1,121 @@
+        TeamMember member = mock(TeamMember.class);
+
+        when(project.getVisibility()).thenReturn(ProjectVisibility.PRIVATE);
+        when(project.getTeams()).thenReturn(List.of(team));
```
тогда и это не надо, если просто сеттерами нужные поля проставить

mipkasnow commented on src/test/java/faang/school/projectservice/validator/project/ProjectValidatorTest.java:
```diff
@@ -0,0 +1,121 @@
+        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class,
+                () -> validator.validateAccessToProject(project, 1L));
+
+        assertEquals("Access denied to private project", ex.getMessage());
```
TestUtils userService

mipkasnow reviewed (commented):
провел ревью. 
Основная проблема - маппер и валидатор не статические, из-за этого лишние зависимости в сервисах и тестировать сложно.
ПО тестам - сущности/дто не надо мокать никогда

Charly-vision committed:
-made changes
-removed unnecessary dependencies
-added default error handling method
-created interface TestUtils for convenience
-implemented Swagger UI

Charly-vision committed:
-Adapted the class ProjectСontroller for Swagger

Charly-vision committed:
-made corrections in class GlobalExceptionHandler and ProjectController

20cognosce commented on src/main/java/faang/school/projectservice/controller/exeption/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,22 @@
+
+    @ExceptionHandler(DuplicateResourceException.class)
+    @ResponseStatus(HttpStatus.CONFLICT)
+    static Map<String, String> handleDuplicateResource(DuplicateResourceException ex) {
```
а нужен ли тут static?

20cognosce commented on src/main/java/faang/school/projectservice/controller/exeption/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,22 @@
+package faang.school.projectservice.controller.exeption;
+
+import faang.school.projectservice.exception.project.DuplicateResourceException;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler(DuplicateResourceException.class)
+    @ResponseStatus(HttpStatus.CONFLICT)
+    static Map<String, String> handleDuplicateResource(DuplicateResourceException ex) {
+        Map<String, String> errorMap = new HashMap<>();
+        errorMap.put("error", ex.getMessage());
+        return errorMap;
+    }
```
Хорошо, но это не масштабируется когда будет +10 ошибок и +10 статус кодов
нужен универсальный метод

20cognosce commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,132 @@
+                    @ApiResponse(responseCode = "400", description = "Invalid input data", content = @Content),
+                    @ApiResponse(responseCode = "409", description = "Project with the same name already exists", content = @Content)
+            }
+    )
```
Это тщательная работа, но на практике особо никто так не заморачивается

Еще большая проблема - это очень перегружает контроллер, за всей документацией не видны сами эндпоинты, очень много лишнего

Прошу вынеси документацию, всё что по сваггеру, в интерфейс
А здесь оставить саму апишку

https://github.com/CorporationX/post_service/pull/2087
Тут пример

20cognosce commented on src/main/java/faang/school/projectservice/mapper/ProjectMapper.java:
```diff
@@ -0,0 +1,46 @@
+                project.getVisibility(),
+                project.getCreatedAt(),
+                project.getUpdatedAt()
+        );
```
@Builder

20cognosce commented on src/main/java/faang/school/projectservice/mapper/ProjectMapper.java:
```diff
@@ -0,0 +1,46 @@
+    public static Project toEntity(ProjectCreateDto dto) {
+        if (dto == null) return null;
+
+        return Project.builder()
+                .name(dto.name())
+                .description(dto.description())
+                .visibility(dto.visibility())
+                .build();
```
@Builder

20cognosce commented on src/test/java/faang/school/projectservice/helpers/TestUtils.java:
```diff
@@ -0,0 +1,18 @@
+            Executable executable) {
+
+        T exception = assertThrows(expectedType, executable);
+        assertEquals(expectedMessage, exception.getMessage());
```
не надо проверять сообщения исключений - немного изменится сообщение, и все тесты упадут(

20cognosce commented on build.gradle.kts:
```diff
@@ -16,6 +16,7 @@ dependencies {
     /**
      * Spring boot starters
      */
+    implementation(group = "org.springdoc", name = "springdoc-openapi-starter-webmvc-ui", version = "2.0.3")
```
старая версия, бери актуальную
может понадобиться спринг бут заодно обновить

20cognosce commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,94 @@
+        log.info("Creating a project by a user {}", ownerId);
+
+        boolean nameExists = projectRepository.existsByOwnerIdAndName(ownerId, projectCreateDto.name());
+        ProjectValidator.validateUniqueProjectNameForOwner(projectCreateDto.name(), ownerId, nameExists);
```
Можно в валидатор передавать `Supplier<Boolean> ` через лямбда-выражение
И в валидаторе делать вызов таким образом

Сделай пожалуйста одинаковый порядок аргументов id, name / name, id

20cognosce commented on src/main/java/faang/school/projectservice/service/project/ProjectService.java:
```diff
@@ -0,0 +1,94 @@
+        return project;
+    }
+
+    public List<ProjectDto> getAllProjects() {
+        return projectRepository.findAll().stream()
+                .map(ProjectMapper::toDto)
+                .toList();
+    }
```
у тебя в фасаде вроде то же самое

20cognosce commented on src/main/java/faang/school/projectservice/validator/project/ProjectValidator.java:
```diff
@@ -0,0 +1,42 @@
+        if (nameExists) {
+            throw new DuplicateResourceException("Project with this name already exists for this user");
+        }
+    }
```
этот метод доработать через Supplier

20cognosce commented on src/main/java/faang/school/projectservice/validator/project/ProjectValidator.java:
```diff
@@ -0,0 +1,42 @@
+        if (project == null) {
+            throw new IllegalArgumentException("Project not found");
+        }
+    }
```
почему публичный метод? нигде за пределами класса не используется

20cognosce commented on src/main/java/faang/school/projectservice/validator/project/ProjectValidator.java:
```diff
@@ -0,0 +1,42 @@
+        if (project.getVisibility() == ProjectVisibility.PRIVATE) {
+            boolean isParticipant = project.getTeams() != null && project.getTeams().stream()
+                    .flatMap(team -> team.getTeamMembers().stream())
+                    .anyMatch(p -> Objects.equals(p.getId(), userId));
```
вот это можно вынеси в какой-нибудь статический метод в сервисе

20cognosce commented on src/main/java/faang/school/projectservice/validator/project/ProjectValidator.java:
```diff
@@ -0,0 +1,42 @@
+        if (status == null && description == null) {
+            throw new IllegalArgumentException("Nothing to update");
+        }
+        validateProjectExists(project);
```
странно выглядит, почему сразу в сервисе ошибку не кидать orElseThrow?

Charly-vision committed:
-moved the documentation for Swagger to a separate class
-made some improvements

20cognosce commented on src/main/java/faang/school/projectservice/controller/exeption/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,60 @@
+
+    @ExceptionHandler(DuplicateResourceException.class)
+    public ResponseEntity<Map<String, Object>> handleDuplicateResource(DuplicateResourceException ex) {
+        log.warn("Duplicate resource: {}", ex.getMessage());
```
почему не log.error?)

20cognosce commented on src/main/java/faang/school/projectservice/controller/project/ProjectApi.java:
```diff
@@ -0,0 +1,96 @@
+
+@Tag(name = "Projects", description = "Endpoints for managing projects")
+@RequestMapping("/api/v1/projects")
+public interface ProjectApi {
```
ProjectControllerApiDocs

20cognosce commented on src/main/java/faang/school/projectservice/controller/project/ProjectController.java:
```diff
@@ -0,0 +1,56 @@
+
+import java.util.List;
+
+@Slf4j
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/projects")
+@Tag(name = "Projects", description = "Endpoints for managing projects")
```
Где порядок

Charly-vision replied:
пирамидка

20cognosce commented on src/main/java/faang/school/projectservice/controller/project/ProjectApi.java:
```diff
@@ -0,0 +1,96 @@
+            @ApiResponse(responseCode = "409", description = "Проект с таким именем уже существует"),
+            @ApiResponse(responseCode = "500", description = "Внутренняя ошибка сервера")
+    })
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
```
Нет, это в контроллере должно быть

20cognosce commented on src/main/java/faang/school/projectservice/controller/project/ProjectApi.java:
```diff
@@ -0,0 +1,96 @@
+import io.swagger.v3.oas.annotations.tags.Tag;
+import jakarta.validation.Valid;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PatchMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseStatus;
```
Все маппинги на эндпоинты в контроллере должны быть видны

20cognosce reviewed (approved):
Контроллер доработай пожалуйста, принято

20cognosce commented:
Выбрана неправильная ветка dev, куда ты пытаешься сделать merge
Нужно в master команды hippogrif12

Пересоздай новый ПР пожалуйста, этот закрываю
